radixdlt-engine/src/main/java/com/radixdlt/atommodel/system/scrypt/SystemConstraintScryptV3.java

```diff
@@ -0,0 +1,817 @@
+			() -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(i, s, r) -> ReducerResult.incomplete(s.prepareStakes(i))
+		));
+		os.procedure(new ShutdownAllProcedure<>(
+			PreparedValidatorUpdate.class, PreparingRakeUpdate.class,
+			() -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(i, s, r) -> ReducerResult.incomplete(s.prepareRakeUpdates(i))
+		));
+		os.procedure(new UpProcedure<>(
+			ResetValidatorUpdate.class, NoValidatorUpdate.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> ReducerResult.incomplete(s.reset(u))
+		));
+		os.procedure(new UpProcedure<>(
+			Staking.class, StakeOwnership.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> ReducerResult.incomplete(s.stake(u))
+		));
+		os.procedure(new UpProcedure<>(
+			UpdatingValidatorStakes.class, ValidatorStakeData.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> ReducerResult.incomplete(s.updateStake(u))
+		));
+
+		os.procedure(new UpProcedure<>(
+			CreatingNextValidatorSet.class, ValidatorBFTData.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> ReducerResult.incomplete(s.nextValidator(u))
+		));
+
+		os.procedure(new UpProcedure<>(
+			CreatingNextValidatorSet.class, EpochData.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> ReducerResult.incomplete(s.nextEpoch(u))
+		));
+
+		os.procedure(new UpProcedure<>(
+			StartingEpochRound.class, RoundData.class,
+			u -> new Authorization(PermissionLevel.SUPER_USER, (r, c) -> { }),
+			(s, u, c, r) -> {
+				if (u.getView() != 0) {
+					throw new ProcedureException("Epoch must start with view 0");
+				}
+
+				return ReducerResult.complete();
+			}
+		));
+	}
+
+	@Override
+	public void main(Loader os) {
+		os.substate(
+			new SubstateDefinition<>(
+				RoundData.class,
+				Set.of(SubstateTypeId.ROUND_DATA.id()),
+				(b, buf) -> {
+					var view = REFieldSerialization.deserializeNonNegativeLong(buf);
+					var timestamp = REFieldSerialization.deserializeNonNegativeLong(buf);
+					return new RoundData(view, timestamp);
+				},
+				(s, buf) -> {
+					buf.put(SubstateTypeId.ROUND_DATA.id());
+					buf.putLong(s.getView());
+					buf.putLong(s.getTimestamp());
+				},
+				p -> p.getView() == 0 && p.getTimestamp() == 0
+			)
+		);
+		os.substate(
+			new SubstateDefinition<>(
+				EpochData.class,
+				Set.of(SubstateTypeId.EPOCH_DATA.id()),
+				(b, buf) -> {
+					var epoch = REFieldSerialization.deserializeNonNegativeLong(buf);
+					return new EpochData(epoch);
+				},
+				(s, buf) -> {
+					buf.put(SubstateTypeId.EPOCH_DATA.id());
+					buf.putLong(s.getEpoch());
+				}
+			)
+		);
+		os.substate(
+			new SubstateDefinition<>(
+				ValidatorStakeData.class,
+				Set.of(SubstateTypeId.STAKE_V1.id(), SubstateTypeId.STAKE_V2.id()),
```

Comment:
Do we have plan to remove `STAKE_V2` on mainnet? If so, do we need an extra byte to store the optional flag, unless every validator must specify a rake percentage by then?